src/VscPreviewExtension/media/buildTree.js

```diff
@@ -0,0 +1,318 @@
+        // Hightlight the circle of chosen node
+        var linenumber = data.split(" ");
+        vis.selectAll("g.node")
+          .select("circle")
+          .style("fill", function (d) {
+            if (isIntersect(parseInt(GetStartLineNumber(d.name)), parseInt(GetEndLineNumber(d.name)), parseInt(linenumber[0]), parseInt(linenumber[1]))) {    //centerNode(d);
+              targetNode = d;
+              return selectedGold
+            } else {
+              return d._children ? "lightsteelblue" : "white";
+            }
+          });
+
+        // Hight the text of chosen node
+        vis.selectAll("g.node")
+          .select("text")
+          .style("fill", function (d) {
+            if (isIntersect(parseInt(GetStartLineNumber(d.name)), parseInt(GetEndLineNumber(d.name)), parseInt(linenumber[0]), parseInt(linenumber[1]))) {    //centerNode(d);
+              targetNode = d;
+              return selectedGold
+            } else {
+              return unSelectedGray;
+            }
+          });
+      })
+    if (targetNode !== currrentCenterNode) {
+      centerNode(targetNode);
+      currrentCenterNode = targetNode;
+    }
+  }, 500);
+
+  function update(source) {
+    var levelWidth = [1];
+    // Count the total number of nodes in every level to compute the height of tree
+    var childCount = function (level, n) {
+      if (n.children && n.children.length > 0) {
+        if (levelWidth.length <= level + 1)
+          levelWidth.push(0);
+        levelWidth[level + 1] += n.children.length;
+        n.children.forEach(function (d) {
+          childCount(level + 1, d);
+        });
+      }
+    };
+    childCount(0, root);
+    var newHeight = d3.max(levelWidth) * 40; // 40 pixels per line  
+    tree = tree.size([newHeight, width]);
+
+    // Compute the new tree layout.
+    var nodes = tree.nodes(root).reverse();
+
+    // Set widths between levels based on maxLabelLength.
+    nodes.forEach(function (d) {
+      d.y = getoffset(d.depth);
+    });
+
+    // Compute the offet of node on the node.depth and maxLabelLengthOfeachLevel
+    function getoffset(level) {
+      var count = 0;
+      for (var i = 1; i <= level; i++) {
+        count += maxLabelLengthOfEachLevel[i] * (22 - i * 2);
+      }
+      return count;
+    }
+
+    // Update the nodes...
+    node = vis.selectAll("g.node")
+      .data(nodes, function (d) { return d.id || (d.id = ++i); });
+
+    // Enter any new nodes at the parent's previous position.
+    var nodeEnter = node.enter().append("svg:g")
+      .attr("class", "node")
+      .attr("transform", function (d) { return "translate(" + source.y0 + "," + source.x0 + ")"; });
+
+    nodeEnter.append("svg:circle")
+      .attr("r", 1e-6)
+      .style("fill", function (d) { return d._children ? "steelblue" : "white"; })
+      .on('click', circleClick);;
+
+    nodeEnter.append("svg:text")
+      .attr("x", function (d) { return d.children || d._children ? -10 : 10; })
+      .attr("dy", ".35em")
```

Comment:
you mean just the `dy` or all of the attributes?
for most of the attributes, their value is dependent on the node's   status but not a constant

---
In reply to: [79999775](https://github.com/dotnet/docfx/pull/693#discussion_r79999775) [](ancestors = 79999775)